File: spring-reactive-01/learn-java/src/main/java/boki/learnjava/mono/MonoEx2.java
package boki.learnjava.mono;

import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Mono;

@Slf4j
public class MonoEx2 {
    public static void main(String[] args) {
        Mono.empty()
            .subscribe(
                data -> log.info("# emitted data: {}", data),
                error -> { },
                () -> log.info("# emitted onComplete signal")
            );

        // 같은 코드
        Mono.empty()
            .doOnNext(data -> log.info("# emitted data: {}", data))
            .doOnError(error -> log.error("Error occurred: {}", error))
            .doOnTerminate(() -> log.info("# emitted onComplete signal"))
            .subscribe();
    }
}

File: learn-kt/src/main/kotlin/boki/learn/mono/MonoEx2.kt
package boki.learn.mono

import reactor.core.publisher.Mono

fun main() {
    Mono.empty<Any>()
        .subscribe(
            { data: Any -> println("# emitted data: $data") },
            { error: Throwable -> {} },
            { println("# emitted onComplete signal") }
        )

    // 같은 코드
    Mono.empty<Any>()
        .doOnNext { data -> println("# emitted data: $data") }
        .doOnError { error -> println("Error occurred: $error") }
        .doOnTerminate { println("# emitted onComplete signal") }
        .subscribe()
}